ISSUE 341: Fix order grid filter error

sethobey opened the pull request:
### Context
<!-- Why is this change necessary? Write one or two sentences to explain what's going on. -->
Closes #341

Customer identified regression in sales order grid when filtering by date. Magento core code does not prefix `created_at` filter value with table alias leading to ambiguous query.

### Description
<!-- What does this PR change? If it's a bug, describe the fix. If it's a feature, post screenshots or a video. -->
- Adds plugin method for Collections to specify `main_table` alias when filtering Order or Creditmemo grids by `created_at` values and OrderCreditmemo grid `order_id`
- Replaces custom UI column components in ui_component XML with base Magento column component definitions to reduce custom code and indirectly enables filtering on TaxJar sync date.

### Performance
<!-- How does this PR impact the area that's being changed? Prove it out. This can be an informal benchmark, EXPLAIN ANALYZE output, etc. -->
Adds overhead of additional type-checking operation when querying for any UI collections, but should still only introduce negligible latency.

### Testing
<!-- How do we test this PR? **This section is critical.** Some ideas:
- Provide clear steps to reproduce w/ test data
- Show us how you tested the PR
- Call out specific areas of concern
-->
#### Order grid
1. Navigate to sales_order_grid
2. Add filter on Purchase Date (`created_at`) and/or Synced To TaxJar (`tj_salestax_sync_date`)
3. Apply filter
4. Observe filter is successfully applied, grid re-loaded

#### Creditmemo grid
1. Navigate to sales_creditmemo_grid
2. Add filter on Created (`created_at`) and/or Synced To TaxJar (`tj_salestax_sync_date`)
3. Apply filter
4. Observe filter is successfully applied, grid re-loaded

#### Sales Order Creditmemo grid
1. Navigate to Sales Order > Creditmemos
2. Add creditmemo filter on Created (`created_at`) and/or Synced To TaxJar (`tj_salestax_sync_date`)
3. Apply filter
4. Observe filter is successfully applied, grid re-loaded

<img width="1360" alt="image" src="https://user-images.githubusercontent.com/47947793/192636259-4990b86a-1584-4be5-bf7e-e51560a69401.png">

#### Versions
<!-- What version(s) did you test this change on? -->
- [X] Magento 2.4
- [ ] Magento 2.3
<!-- What edition(s) of Magento did you test this change on? -->
- [X] Magento Open Source (formerly Magento 2 Community Edition)
- [ ] Adobe Commerce (formerly Magento 2 Enterprise Edition)
<!-- What version of PHP did you test this change on? -->
- [ ] PHP 8.2
- [X] PHP 8.1
- [ ] PHP 7.4
- [ ] PHP 7.3

sethobey-stripe committed:
Fixes sales order and creditmemo grid filtering ambiguous query

dallendalton reviewed (commented):
Should we also remove any unnecessary code from AddTjSyncDateToGrid.php. Specifically are the addFilterToMap method calls still necessary?

sethobey-stripe committed:
Refactors AddTjSyncDateToGrid

sethobey-stripe committed:
Removes custom UI column classes

sethobey commented:
> Should we also remove any unnecessary code from AddTjSyncDateToGrid.php. Specifically are the addFilterToMap method calls still necessary?

@dallendalton Good call out!

In consideration of your comment I refactored the `AddTjSyncDateToGrid::class` to remove duplicated code, then wrote some integration tests to validate the joins. In the process, I identified another potential ambiguous query (`order_id`) when filtering the sales-order-creditmemo grid which is also now addressed in `BeforeApplyFilters::class` plugin.

While I was at it, I updated our sales-related `ui_component` XML overrides to use Magento's default date column types, allowing us to remove 3 (unnecessary) custom UI component column definitions. This also now enables filtering on TJ sync date values.

sethobey-stripe committed:
Clean up

sethobey-stripe committed:
Adds test coverage for BeforeApplyFilters plugin
